src/transport/http_low_level.rs

```diff
@@ -0,0 +1,195 @@
+use async_trait::async_trait;
+
+#[cfg(not(target_arch = "wasm32"))]
+pub(crate) type HyperClient<C> = hyper::Client<C, hyper::Body>;
+
+#[cfg(all(feature = "tokio-rt", not(target_arch = "wasm32")))]
+pub(crate) type HttpsConnector = hyper_tls::HttpsConnector<hyper::client::HttpConnector>;
+
+#[cfg(all(feature = "tokio-rt", not(target_arch = "wasm32")))]
+pub(crate) type HyperHttpsClient = HyperClient<HttpsConnector>;
+
+#[cfg(target_arch = "wasm32")]
+pub(crate) struct WasmClient;
+
+/// A low level HTTP trait.
+#[async_trait(?Send)]
+pub trait HttpLowLevel {
+    /// HTTP Method type.
+    type Method;
+
+    /// HTTP Header map type.
+    type HeaderMap;
+
+    /// HTTP Body type.
+    type Body;
+
+    /// HTTP Response type.
+    type Response;
+
+    /// Send a request and receive a response.
+    async fn send(
+        &self,
+        method: Self::Method,
+        uri: &str,
+        headers: Self::HeaderMap,
+        body: Self::Body,
+    ) -> crate::Result<Self::Response>;
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+#[async_trait(?Send)]
+impl<C> HttpLowLevel for HyperClient<C>
+where
+    C: hyper::client::connect::Connect + Clone + Send + Sync + 'static,
+{
+    type Method = http::Method;
+    type HeaderMap = http::HeaderMap;
+    type Body = Vec<u8>;
+    type Response = http::Response<hyper::Body>;
+
+    async fn send(
+        &self,
+        method: Self::Method,
+        uri: &str,
+        headers: Self::HeaderMap,
+        body: Self::Body,
+    ) -> crate::Result<Self::Response> {
+        let mut builder = http::Request::builder().method(method).uri(uri);
+        if let Some(h) = builder.headers_mut() {
+            *h = headers;
+        }
+
+        let request = builder.body(body.into()).unwrap();
```

Comment:
Here we could `map_err(|err| typesense::Error::TypesenseClientError)?` what do you think? Maybe we could at least associate TypesenseClientError with a String so we could know more information about the error. What do you think?

The same for the other unwraps on this file.

Reply:
Sure we could do that, the only thing we would miss is for different errors we would get the same type back i.e. TypesenseClientError.